src/utils/daos_hdlr.c

```diff
@@ -1568,11 +1600,1359 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	char *name = NULL, *dname = NULL;
+
+	if (src_file_dfs.type == POSIX && dst_file_dfs.type == DAOS) {
+		parse_filename_dfs(src_str, &name, &dname);
+		mode_t tmp_mode_dir = S_IRWXU;
+
+		/* construct destination directory in DAOS, this needs
+		 * to strip the dirname and only use the basename that is
+		 * specified in the dst argument
+		 */
+		char dst_dir[MAX_FILENAME];
+
+		src_str_len = strlen(dname);
+		int path_length = snprintf(dst_dir, MAX_FILENAME, "%s/%s",
+				       dst_str, src_str + src_str_len);
+
+		if (path_length >= MAX_FILENAME) {
+			rc = ENAMETOOLONG;
+			fprintf(stderr, "Path length is too long.\n");
+			D_GOTO(out_disconnect, rc);
+		}
+		rc = file_mkdir(&dst_file_dfs, dst_dir, &tmp_mode_dir);
+		if (rc != EEXIST && rc != 0)
+			D_GOTO(out_disconnect, rc);
+		rc = fs_copy(&src_file_dfs, &dst_file_dfs,
+			     src_str, src_str_len, dst_str);
+		if (rc != 0)
+			D_GOTO(out_disconnect, rc);
+	} else if (src_file_dfs.type == DAOS && dst_file_dfs.type == POSIX) {
+		mode_t tmp_mode_dir = S_IRWXU;
+
+		rc = file_mkdir(&dst_file_dfs, dst_str, &tmp_mode_dir);
+		if (rc != EEXIST && rc != 0)
+			D_GOTO(out_disconnect, rc);
+		rc = fs_copy(&src_file_dfs, &dst_file_dfs,
+			     src_str, src_str_len, dst_str);
+		if (rc != 0)
+			D_GOTO(out_disconnect, rc);
+	} else if (src_file_dfs.type == DAOS && dst_file_dfs.type == DAOS) {
+		parse_filename_dfs(src_str, &name, &dname);
+		mode_t tmp_mode_dir = S_IRWXU;
+		/* construct destination directory in DAOS, this needs
+		 * to strip the dirname and only use the basename that is
+		 * specified in the dst argument
+		 */
+		char dst_dir[MAX_FILENAME];
+		int path_length = snprintf(dst_dir, MAX_FILENAME, "%s/%s",
+				       dst_str, src_str + src_str_len);
+		if (path_length >= MAX_FILENAME) {
+			rc = ENAMETOOLONG;
+			fprintf(stderr, "Path length is too long.\n");
+			D_GOTO(out_disconnect, rc);
+		}
+		rc = file_mkdir(&dst_file_dfs, dst_dir, &tmp_mode_dir);
+		if (rc != EEXIST && rc != 0)
+			D_GOTO(out_disconnect, rc);
+		rc = fs_copy(&src_file_dfs, &dst_file_dfs,
+			     src_str, src_str_len, dst_str);
+		if (rc != 0)
+			D_GOTO(out_disconnect, rc);
+	/* TODO: handle POSIX->POSIX case here */
+	} else {
+		fprintf(stderr, "Regular POSIX to POSIX copies are not "
+			"supported\n");
+		D_GOTO(out_disconnect, rc = EINVAL);
+	}
+
+	if (dst_file_dfs.type == DAOS) {
+		fprintf(stdout, "Successfully copied to DAOS: %s\n",
+			dst_str);
+	} else if (dst_file_dfs.type == POSIX) {
+		fprintf(stdout, "Successfully copied to POSIX: %s\n",
+			dst_str);
+	}
+out_disconnect:
+	/* umount dfs, close conts, and disconnect pools */
+	rc = fs_copy_disconnect(&fa, &src_file_dfs, &dst_file_dfs);
+	if (rc != 0)
+		fprintf(stderr, "failed to disconnect (%d)\n", rc);
+out:
+	if (src_file_dfs.type == DAOS) {
+		if (src_str != NULL)
+			free(src_str);
+	}
+	if (dst_file_dfs.type == DAOS) {
+		if (dst_str != NULL)
+			free(dst_str);
+	}
+	return rc;
+}
```

Comment:
Quite a lot of code here, should probably be moved to a separate file eventually.

Reply:
@johannlombardi I think after the dfs wrapper functions are moved that should reduce the amount of code added to daos_hdlr.c significantly. Link to that is here: https://jira.hpdd.intel.com/browse/DAOS-6098